features/publish.feature

```diff
@@ -0,0 +1,131 @@
+      | meta             |
+      | source           |
+      | gherkinDocument  |
+      | pickle           |
+      | stepDefinition   |
+      | testRunStarted   |
+      | testCase         |
+      | testCaseStarted  |
+      | testStepStarted  |
+      | testStepFinished |
+      | testCaseFinished |
+      | testRunFinished  |
+
+  @spawn
+  Scenario: Report is published when CUCUMBER_PUBLISH_TOKEN is set
+    Given a report server is running on 'http://localhost:9987'
+    When I run cucumber-js with arguments `` and env `CUCUMBER_PUBLISH_TOKEN=keyboardcat CUCUMBER_PUBLISH_URL=http://localhost:9987/api/reports`
+    Then it passes
+    And the server should receive the following message types:
+      | meta             |
+      | source           |
+      | gherkinDocument  |
+      | pickle           |
+      | stepDefinition   |
+      | testRunStarted   |
+      | testCase         |
+      | testCaseStarted  |
+      | testStepStarted  |
+      | testStepFinished |
+      | testCaseFinished |
+      | testRunFinished  |
+    And the server should receive an "Authorization" header with value "Bearer keyboardcat"
+
+  @spawn
+  Scenario: a banner is displayed after publication
+    Given a report server is running on 'http://localhost:9987'
+    When I run cucumber-js with arguments `--publish` and env `CUCUMBER_PUBLISH_URL=http://localhost:9987/api/reports`
+    Then the error output contains the text:
+      """
+      ┌──────────────────────────────────────────────────────────────────────────┐
+      │ View your Cucumber Report at:                                            │
+      │ https://reports.cucumber.io/reports/f318d9ec-5a3d-4727-adec-bd7b69e2edd3 │
+      │                                                                          │
+      │ This report will self-destruct in 24h unless it is claimed or deleted.   │
+      └──────────────────────────────────────────────────────────────────────────┘
+      """
+
+  @spawn
+  Scenario: when results are not published, a banner explains how to publish
+    When I run cucumber-js
+    Then the error output contains the text:
+      """
+      ┌──────────────────────────────────────────────────────────────────────────┐
+      │ Share your Cucumber Report with your team at https://reports.cucumber.io │
+      │                                                                          │
+      │ Command line option:    --publish                                        │
+      │ Environment variable:   CUCUMBER_PUBLISH_ENABLED=true                    │
+      │                                                                          │
+      │ More information at https://reports.cucumber.io/docs/cucumber-js         │
+      │                                                                          │
+      │ To disable this message, add this to your ./cucumber.js:                 │
+      │ module.exports = { default: '--publish-quiet' }                          │
```

Comment:
I feel conflicted about this. This feels annoying to me as a user. Certainly glad we implemented the CLI option to suppress it but why does this need to be outputted a user runs the tool without publish?

Could this be a message shown on install / the help page / somewhere visible on the README?

Reply:
We've already implemented it this way in Cucumber-Ruby and Cucumber-JVM. So far we've not heard any complaints, just positive feedback.

We want as many people as possible to discover this new functionality, and showing it after the run will reach more people than if it's only shown on install and in documentation (which many people never look at).

Let's see if anyone complains.

Reply:
Me. I'll complain.  I'm beginning the setup of a test suite and this message is annoying to look at.  I tried dropping the module.exports into my step definition file but it didn't get rid of this message. So how do I get rid of this?

Reply:
> I tried dropping the module.exports into my step definition file

You're not supposed to drop that in your step definition files, but in your global `cucumber.js` [config file](https://github.com/cucumber/cucumber-js/blob/master/docs/cli.md#profiles) at the root of your project.

How can we make this harder to misunderstand @michaelm-rsi?

> I'm working on a corporate project which will NEVER publish test results onto a public server for security reasons.

I completely understand that, which is why we've added an option to turn this off. We're also considering making Cucumber Reports available as a server you can install on-prem.